### ISSUE TITLE
Entity recognition (continued)

### ISSUE THREAD

**Issue by dtrieschnigg** [marcotcr/checklist#78]
Related issue  #41 

Thank you for creating this, it's a really insightful tool to thoroughly test the output of an NLP system! Today we tried to use Checklist to evaluate a named entity recognition system. This system outputs stand-off annotation to indicate where entities are found. We extended the work you did [here](https://github.com/marcotcr/checklist/blob/master/notebooks/Testing%20NER%20(preliminary).ipynb), by adding the offsets of generated entities to the `meta` information.
This works fine as long as we stick to the template.

However, we would also like to generate cases which slightly change the template (for instance by deleting, changing or inserting characters), but this messes up the offsets required for evaluation.

Do you have any plans to support evaluation of such systems?

**Comment by marcotcr** [comment]
> Do you have any plans to support evaluation of such systems?    

Yes, I would like to support token-based predictors. 
I can try to add some helpers to make token based expectation functions easier to write, but I need to spend some time thinking and looking through examples of tests that people would want to write.

Can you give a few examples of tests you want to write but are having a hard time right now? It seems like you're trying to write perturbation tests, but I'm not sure if I understood it quite well. If you can, please include examples of:
- The output of your NER system
- A few examples within a test case (e.g. if you're doing perturbations, you could do 'John is a nice man' -> 'Luke is a nice man')
- What the expectation function should be (in words)

Thanks!

**Comment by dtrieschnigg** [comment]
> The output of your NER system

Example input: `John is a nice man`
Output from the NER system: `<Name span=(0,4)>`

What we did was create a template such as '{name} is a nice man`
With a list of possible names, e.g. `['John', 'Joe']`
This should result in expected output: `<Name span[0,4]>` for John and `<Name span[0,3]>` for Joe.
Currently we put the expected output in the meta information so the expectation function can easily check that with was actually produced by the NER system.

If we would like to combine this random character perturbations (including insertions and deletions), this screws up the offsets. For instance when randomly deleting a character from `John is a nice man`, it could become `Jhn is a nice man`. The NER system should return `(0,3)` rather than `(0,4)`. But we can't easily check what the perturbations actually changed so it's hard to update the offsets accordingly (especially when the inserted name has been changed).

So for NER we are interested in perturbations such as:
- case
- punctuation
- spelling errors

So in short, we are looking for a way to keep track of generated offset-based (doesn't matter whether it is token or character-based) annotations. Trying to reproduce the offsets by searching in the generated text feels a big clunky (and become difficult with spelling errors and changes in punctuation).

> What the expectation function should be (in words)
Checking whether the generated named entity offsets match the one produced by the NER. (a simple implementation would only return true iff all entities exactly match the entities found by NER).

**Comment by marcotcr** [comment]
Hm, I see two ways of doing what you want right now.
1. Write an expectation function that takes `(x1, x2, meta)` and updates the `meta` based on the diff between `x1` and `x2`. 
It's actually not very difficult to track the offsets, you just have to use something like the code below to get the changes and update the expected spans, e.g. 'if I add something in position 1, every span that starts to the right of 1 gets updated by `+1`', 'if I delete something within a span, the number on the right of the span gets reduced by 1'
```python
matcher = difflib.SequenceMatcher(None, x1, x2)
for tag, i1, i2, j1, j2 in matcher.get_opcodes():
    if tag == 'equal':
        continue
    print(tag, i1, i2, j1, j2)
```

2. Write the new expected span in the perturbation function.
Let's say you have some `t` you generated with the editor (such that you have `t.data` and `t.meta`), or some original data that you ran your NER system on. You can do something like the following:
```python
from checklist.perturb import Perturb
def perturb_ner(example, meta=True):
  sentence, ner_annotations = example
  # perturb sentence here, update the ner annotation
  return [sentence, new_sentence], [ner_annotations, new_meta]
pairs = list(zip(t.data, t.meta)) # or pairs = zip(data, ner_output)
t = Perturb.perturb(pairs, perturb_ner, meta=True, keep_original=False)
```

I don't know how I could write a helper to make this easier, except maybe if I wrote something that helps write token-based expectations and a way for perturbation functions to keep track of token alignment between source and perturbed sentences. 

I'll think some more about it, but does this make sense? Have you tried one of the options above?

**Comment by dtrieschnigg** [comment]
The second approach looks cleaner: keeping the generation and modification of text and annotations in one place. The first approach feels a bit like reverse engineering which might fail for complex cases (for instance when you shuffle the words in a sentence)
I think the cleanest option would be to have a string-like structure which keeps track of text and (offset) annotations. When you edit the text, the annotations change with it.

**Comment by marcotcr** [comment]
Closing this for now. Thanks for the discussion
